x-pack/plugin/vectors/src/main/java/org/elasticsearch/xpack/vectors/models/LSHModelEuclideanMultiProbe.java

```diff
@@ -0,0 +1,360 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+
+package org.elasticsearch.xpack.vectors.models;
+
+import org.apache.lucene.util.BytesRef;
+import org.apache.lucene.util.InPlaceMergeSorter;
+import org.elasticsearch.common.Randomness;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.List;
+import java.util.PriorityQueue;
+import java.util.Random;
+
+/**
+ * Implementation of the multi-probe LSH based on the papers:
+ *  1. Datar, Mayur, et al. "Locality-sensitive hashing scheme based on p-stable distributions."
+ *  Proceedings of the twentieth annual symposium on Computational geometry. ACM, 2004.
+ *  2. Lv, Qin, et al. "Multi-probe LSH: efficient indexing for high-dimensional similarity search."
+ *  Proceedings of the 33rd international conference on Very large data bases. VLDB Endowment, 2007.
+ *
+ *  Given the user's provided parameters: D (number of dims), L(number of hash tables),
+ *  K(number of hash functions in each hash table) and W (with of the projection), we:
+ *  - generate L * K number of random a and b parameters for the L * K number of hash functions in the form (a * X + b)/W.
+ *      Each hash function hashes a vector to integer.
+ *  - compute pertSets — sets of indexes in increasing order of their scores precomputed based on the scores' expectations
+ *  Thus, the model contains and stores the following numbers: D, L, K, W, array for a, array for b, and array for pertSets.
+ *
+ *  Hashing a document vector:
+ *  - each hash function hashes a vector to an integer
+ *  - to hash a vector, we combine results from K hash functions to get a byte array of K integers.
+ *      Increasing K increases precision, as for 2 vectors to match, all their K integers must match.
+ *  - we hash each vector this way L times (each time using different K hash functions). This corresponds to L hash tables.
+ *      We add to the vector's K integer bytes another 2 bytes to encode the hash table number (0,1,..L-1) and get a BytesRef from it.
+ *      Increasing L increases recall. During matching we use Bool SHOULD condition for each of L hashes.
+ *  - the resulted L hashes(BytesRef) is how the vector will be indexed into a Lucene field.
+ *
+ *  Hashing a query vector:
+ *  - a query vector is hashed like a document vector: L number of hashes
+ *      plus L * (numProbes - 1) number of other hashes.
+ *
+ */
+
+public class LSHModelEuclideanMultiProbe {
+    private static final byte MAX_PERT_SET_LENGTH = 10; // max numb of neighboring hash cells in each hash table to consider as candidates
```

Comment:
not sure about the max limit for `MAX_PERT_SET_LENGTH`. For the audio set I used `20` with `number_of_probes=20` for a query.